### PR TITLE
STM32: Enhance PWM driver

### LINKED ISSUE #6625: stm32: pwm: PWM 3 typo
pwm_stm32_2 should be pwm_stm32_3

https://github.com/zephyrproject-rtos/zephyr/blob/9d2915b763c9ead8478099bd84766673ca6162aa/drivers/pwm/pwm_stm32.c#L251

### PR DESCRIPTION
**Pull request by vaussard**
This patchset performs the following tasks:

- Ease device registration by using a macro to factorize common code
- Add support for more PWM devices
- Replace hardcoded prescalers by Kconfigs

Fixes #6625

### CODE REVIEW

**Review by erwango — CHANGES_REQUESTED** [review]
LGTM for user selectable prescaler.

Code concatenation should be reworked to take F0 into account.

**Review by mbolivar — COMMENTED** [review]
Approach in the last 3 patches looks sane -- seems like this needs a rebase though

**Comment by vaussard** [comment]
Hi @mbolivar and @erwango, thanks for the review!

@erwango : The current driver does not seem to have F0-specific bits. Do you want me to port this driver to F0? I do not have such a board around to test. I can add the necessary part if you tell me what to look for. Is the clock name different on F0 ?

@mbolivar : Rebase done.

**Comment by erwango** [comment]
@vaussard : You don't need to develop PWM for F0, but considering this is an exception to the rule you used to factorize code. I'd like it to be taken into account from the start.
F0: `#define LL_APB1_GRP2_PERIPH_TIM1` (same for 15/16/17)
F1/F3/...: `#define LL_APB2_GRP1_PERIPH_TIM1` (same for 8/9/10/11/15/16/17)

**Comment by vaussard** [comment]
Ok @erwango, so I will add a new patch to support F0 and see if nucleo_f091rc and alike compile. I won't be able to test however. Would it be acceptable for you?

**Comment by erwango** [comment]
@vaussard : Yes, of course. You can mention in your commit message that minimum is done on F0 to compile but feature should not be considered as available on F0 until futher dev/test.

**Comment by vaussard** [comment]
Hi @erwango, I addressed your comments (and even more):

- Added all the PWM up to 20 as I realized that some MCUs can have them
- Added the STM32F0-specific clocks
- Fixed the compilation of the PWM driver on STMF3 and nucleo_f334r8

**Review by dbkinder — DISMISSED** [review]
+1 for docs

**Comment by vaussard** [comment]
I just rebased due to forced changes in the arm branch.
Sorry @dbkinder could you tell me where you would like to see more docs? Github does not allow me anymore to see the location of your comment due to the rebase.

**Review by ydamigos — COMMENTED** [review]
Could you move commit 8caa2cf3ffbe7c0229b3baed1fc0fd076281fc5f before commit 1aedb4e54e669df869e8cc0c31485b6542c4deb8 and squash commits 1aedb4e54e669df869e8cc0c31485b6542c4deb8 with bc0db4de0ea839431dd713a266dc2cda7f2a2cd0 and aad0b8dc0ca2d8f73f6aedb89d285cd92517ceb7 with 8caa2cf3ffbe7c0229b3baed1fc0fd076281fc5f?

**Comment by vaussard** [comment]
Hi @ydamigos. These patches were done later and the only reason that I did not squashed 1aedb4e and bc0db4d was to avoid facing a huge amount of conflicts. But I will do that if you insist :-)
But I am not sure about moving 8caa2cf before commit 1aedb4e, because it is more related to the STM32F0 changes that are done in bde95a9.

**Comment by ydamigos** [comment]
@vaussard My comments are not blocking. IMO, the proposed changes make the PR easier to check.

>But I am not sure about moving 8caa2cf before commit 1aedb4e, because it is more related to the STM32F0 changes that are done in bde95a9.

In that case, I would move both 8caa2cf and bde95a9 before 1aedb4e.

**Comment by vaussard** [comment]
@ydamigos Moving 8caa2cf and bde95a9 before 1aedb4e will force me to rewrite almost all the patches due to the conflicts. IMHO such ordering would be acceptable:

1. Convert to use a macro
2. Add all PWM up to PWM20
3. Use Kconfig for prescalers
4. Add clock group
5. Add STM32F0 bits

It boils down to enhancing-before-extending or extending-before-enhancing. I am OK with both but one is more work to rewrite.

**Comment by ydamigos** [comment]
> Convert to use a macro
    Add all PWM up to PWM20
    Use Kconfig for prescalers
    Add clock group
    Add STM32F0 bits

@vaussard Sounds good.

**Comment by dbkinder** [comment]
Not sure now what docs were updated that I reviewed before the rebase... the API material will update automatically if the header files changed (http://docs.zephyrproject.org/api/io_interfaces.html#pwm-interface).  Might be worth looking through the PWM samples docs (http://docs.zephyrproject.org/samples/basic/basic.html)  to see if they're still OK or if a new sample would be useful.

**Comment by vaussard** [comment]
Hi @ydamigos, I rebased as discussed, let me know if this is OK for you.

**Comment by vaussard** [comment]
Hi @dbkinder. The API is still the same, I only added a few more devices and a Kconfig to set the prescaler. I compiled the PWM sample without changes.

**Comment by vaussard** [comment]
Hi @mbolivar and @erwango, I did addressed your comments. Maybe could you have a look ? Thanks.

**Review by ydamigos — APPROVED** [review]

File: drivers/pwm/pwm_stm32.c
Comment (erwango) [inline drivers/pwm/pwm_stm32.c]: Is that possible to keep "TIM_" prefix in the timer argument?
"1" is not quite meaningful."TIM_1" would help reader to know what this is about without need to go back to the macro definition.

**Review by erwango — CHANGES_REQUESTED** [review]
One last request otherwise looks good!

File: drivers/pwm/pwm_stm32.c
Comment (vaussard) [reply drivers/pwm/pwm_stm32.c]: Ok this looks reasonable. I will do it.

Comment (vaussard) [reply drivers/pwm/pwm_stm32.c]: Mmmh but there is an issue with your proposal. If I replace "1" by "TIM1", this will be used at other places in the maro, like naming the cfg and data structures (that's ok I guess) and the DEV_NAME Kconfig. Thus CONFIG_PWM_STM32_1_DEV_NAME will be renamed to CONFIG_PWM_STM32_TIM1_DEV_NAME. I am not sure if this is desirable.

A workaround would be to have 2 parameters to the macro (like `PWM_DEVICE_INIT_STM32(TIM1, 1, ...)`) but this is not very effective.

Comment (erwango) [reply drivers/pwm/pwm_stm32.c]: Ok, nevermind then, thanks for having had a look.

**Review by erwango — APPROVED** [review]

**Comment by mbolivar** [comment]
Hi @vaussard, I was just pointing out it needed a rebase, I wasn't trying to review the series

**Review by galak — DISMISSED** [review]
Rather than introduce 20 PWM KConfig vars, etc lets do this via device tree.

1. Lets introduce binding and dt nodes for PWM on STM32
2. As part of binding we can get the prescalar set in the DT and generate that

**Review by dbkinder — COMMENTED** [review]
no docs to review

**Comment by galak** [comment]
Any update on this?

**Comment by vaussard** [comment]
Hi @galak. No sorry, I am trying to find some spare time to address your comments. Hope to do that in the near future.

**Comment by vaussard** [comment]
Hi @galak, I addressed your suggestions by adding a PWM binding and the corresponding pwm nodes in all device trees.
I also fixed another issue while compiling for STM32F0, see commit c808526.

**Review by erwango — CHANGES_REQUESTED** [review]
Hi @vaussard ,

Some comments on device tree introduction for STM32:
For stm32 device tree, we target to stay close to Linux dts files (with the medium term target to  us a single repo between zephyr, Linux and other projects using device tree on stm32). To do this we should stay closest to device tree description used for stm32 in Linux.
cf: https://elixir.bootlin.com/linux/v4.16-rc2/source/arch/arm/boot/dts/stm32f429.dtsi
For PWM, you could see that PWM are sub-nodes of timers (PMW is particular use of timer IP). As far as possible, we should stay close to this model.
Then, please enable pwm in dts only for boards in which have PWM=y.

Last, sorry to come with this late, but prescaler could be computed from the timer clock value and required period, no need to provide it as a parameter.

**Comment by vaussard** [comment]
Hi @erwango 

1. I looked into the Linux binding before doing the Zephyr's binding, but I was in doubt if such subnode would be supported by the Python parser. Ok I will test...
2. I enabled in the dts for all boards with a reference to PWM in their pinmux. The user is then free to enable the driver or not. Only 7 boards do have CONFIG_PWM=y in their defconfig, but 14 boards have a pinmux related to PWM. What do you think?
3. I also thought about computing the prescaler when setting the period, but I regard this as a separate issue. This pull request has already grown quite a lot since the original submission.

Thanks

**Comment by vaussard** [comment]
Hi @erwango, I did a quick test for the new DTS binding. See the 4 WIP patches for STM32L4. Of course I will address other STM32 families and properly rebase once the details are ironed out.

It does not look too bad, but I had to use the "bus" property in the .yaml binding in order to generate defines that could be compiled. Otherwise I got something like

`#define ST_STM32_PWM_40000000/PWM_LABEL                 "PWM_2"`

I am not a big fan, because it is not a real bus, only a child node. As you are knowledgeable about `extract_dts_includes.py`, do you know any better way?

**Comment by erwango** [comment]
Hi  @vaussard,

You can try following branch to fix node name generation issue.
https://github.com/erwango/zephyr/tree/fix_node_generation

Otherwise this goes in good direction!
Missing information I see from dts is PWM channel.
I had a look to linux to Documentation/devicetree/bindings/pwm/pwm-stm32.txt,
and it does not look quite clear on this purpose. 
Hence I guess the general Documentation/devicetree/bindings/pwm/pwm.txt applies.

**Comment by vaussard** [comment]
Hi @erwango,

I used your commit "scripts: extract_dts_includes: fix node name when it includes "@"" and it fixes the issue that I had. So I cherry-picked it into this branch as I now depend on it. Let me know if you prefer another solution.

I changed the binding for all STM32 families and I fixed all the patches accordingly. I also rebased on latest master branch to fix conflicts.

Regarding the PWM channels, do you mean the "#pwm-cells" property for the PWM controller nodes? I am not sure if it is needed here, because each controller provides exactly 1 channel. And it is not used in the STM32 Linux device trees (see arch/arm/boot/dts/stm32f429.dtsi for example).

@galak and @erwango could you review the latest changes? Many thanks.

**Comment by erwango** [comment]
Hi @vaussard 

> I used your commit "scripts: extract_dts_includes: fix node name when it includes "@"" and it fixes the issue that I had. So I cherry-picked it into this branch as I now depend on it. Let me know if you prefer another solution.

This is fine

> I changed the binding for all STM32 families and I fixed all the patches accordingly. I also rebased on latest master branch to fix conflicts.

Ok, thanks

> Regarding the PWM channels, do you mean the "#pwm-cells" property for the PWM controller nodes? I am not sure if it is needed here, because each controller provides exactly 1 channel. And it is not used in the STM32 Linux device trees (see arch/arm/boot/dts/stm32f429.dtsi for example).

Ack

About prescaler, I still think that it should be dynamically computed. This approach has been used in mBed and I recommend to use it right now, since it is the logical evolution and it is a lot lines here.

Otherwise looks good

**Comment by vaussard** [comment]
Hi @erwango,

> About prescaler, I still think that it should be dynamically computed. This approach has been used in mBed and I recommend to use it right now, since it is the logical evolution and it is a lot lines here.

So I briefly looked into this topic. Unfortunately I believe that there is a problem with the PWM API that prevent us from doing so. If you look at `pwm_pin_set_usec()` in `include/pwm.h`:
- `pwm_get_cycles_per_sec()` is called to get the number of timer cycles per second
- Then the requested PWM period is converted into hardware cycles, before calling `pwm_pin_set_cycles()`

Problem: if we compute the prescaler in `pwm_pin_set_cycles()`, this will change the number of cycles per second returned by `pwm_get_cycles_per_sec()`. As a result, `pwm_pin_set_usec()` cannot work. The problem is `pwm_pin_set_cycles()` taking hardware cycles as parameters, which makes it unworkable to change the prescaler on the fly.

IMHO it is not possible to have dynamic prescalers with the current PWM API.

**Comment by vaussard** [comment]
Rebased on master to remove conflicts.

**Comment by erwango** [comment]
Hi @vaussard , sorry for delay in review, I'll try to provide feedback in coming days

**Comment by erwango** [comment]
Hi @vaussard ,

You're correct, API prevents from doing dynamic prescaler computation.
This is a shame since pwm_get_cycles_per_sec which prevents doesn't seem to be used at all,
and hence could be removed so that cycles computation is left to the driver.
I'm raising an issue in that direction on PWM API.

So forget about this query for now, will accept your change after rebase.

**Comment by erwango** [comment]
Created #6958

**Comment by vaussard** [comment]
Hi ·@erwango,

The problem is not only with pwm_get_cycles_per_sec() but also with pwm_pin_set_cycles() which takes hardware cycles as argument and is used only in tests. These 2 APIs should be thrown away to keep only pwm_pin_set_usec(). But yeah, let's resolve that separately.

I rebased on master. I had to edit your patch for extract_dts_includes.py as the code has been moved to globals.py, but the change is identical.

**Comment by vaussard** [comment]
Ok so `tests/net/utils/` is failing to build for `nucleo_f103rb` because 20488 bytes of SRAM are used (20 kB total). If I disable the PWM, 20404 bytes are used and it builds. But this looks marginally related to my PR, as this test puts a big pressure on the SRAM. I can disable the PWM, but this test will break next time that someone adds a new member in `struct foo` somewhere else.

**Comment by ydamigos** [comment]
I faced the same problem on olimexino_stm32 board. I disable I2C and SPI to allow the test to build (PR #6873). I followed this solution because, according to issue #6858, we will only configure
board valuable HW and connectors on board level and the user should enable what is needed on application level.

**Comment by vaussard** [comment]
Thank you @ydamigos, I did the same for the PWM on nucleo_f103rb.
So @erwango are we good now? Thanks.

File: scripts/dts/extract/globals.py
Comment (erwango) [inline scripts/dts/extract/globals.py]: Following recent merge of ba2ce2e9f0dcccef8b14c0aea2bf5e483ee07f82, remove the call to ``.upper()``

**Review by erwango — CHANGES_REQUESTED** [review]
One  last comment to take into account following rebase, otherwise, I'm good.

**Review by erwango — APPROVED** [review]

**Comment by vaussard** [comment]
Hi @galak, it looks like you are the last ACK that I need to merge this work. I addressed your request regarding converting the driver to DTS. Could you have a look? Many thanks.

There are again conflicts, but I will address them just before merging to avoid endlessly rebasing this patchset.

**Comment by erwango** [comment]
>  Rather than introduce 20 PWM KConfig vars, etc lets do this via device tree.
    Lets introduce binding and dt nodes for PWM on STM32
    As part of binding we can get the prescalar set in the DT and generate that

@galak : Can you have a new check so we can get this in?

**Comment by vaussard** [comment]
Ping (2) ? This PR was approved by @erwango more than a month ago. I know that people are overloaded with PR reviews, but this PR looks ready to be merged.

The list of conflicts is growing but I will not rebase until I get the confirmation that we are ready to merge as I already rebased numerous times in the past.

**Comment by erwango** [comment]
@vaussard , can you use SDK v0.9.3 when you rebase?
It will generate some warnings you'd need to address:
```
nucleo_f401re.dts_compiled: Warning (avoid_unnecessary_addr_size): unnecessary #address-cells/#size-cells without "ranges" or child "reg" property in /soc/timers@40010000
nucleo_f401re.dts_compiled: Warning (avoid_unnecessary_addr_size): unnecessary #address-cells/#size-cells without "ranges" or child "reg" property in /soc/timers@40000000
```

**Comment by vaussard** [comment]
Hi @erwango

> can you use SDK v0.9.3 when you rebase?

Sure I will update my SDK.

**Comment by vaussard** [comment]
Hi @erwango, I rebased and fixed DTS warnings.
The CI is failing to build some tests, but this does not look to be my fault as it involves `spi_handlers.c`:

```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/spi/spi_handlers.c:32:18: error: implicit declaration of function ‘memcpy’ [-Werror=implicit-function-declaration]
  bufs->buffers = memcpy(buf_copy,
                  ^~~~~~
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/spi/spi_handlers.c:32:16: error: assignment makes pointer from integer without a cast [-Werror=int-conversion]
  bufs->buffers = memcpy(buf_copy,
                ^
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/spi/spi_handlers.c:40:25: error: initialization discards ‘const’ qualifier from pointer target type [-Werror=discarded-qualifiers]
   struct spi_buf *buf = &bufs->buffers[i];
                         ^
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/spi/spi_handlers.c: In function ‘_handler_spi_transceive’:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/spi/spi_handlers.c:86:3: error: implicit declaration of function ‘memset’ [-Werror=implicit-function-declaration]
   memset(&tx_bufs_copy, 0, sizeof(tx_bufs_copy));
   ^~~~~~
cc1: all warnings being treated as errors
```

I took the master branch from a few hours ago, strange.

**Comment by vaussard** [comment]
I cannot reproduce here, `tests/kernel/mem_protect/userspace/` builds fine on `disco_l475_iot1`. And even stranger, if I look into `drivers/spi/spi_handlers.c` at the tip of my tree (1b5add1d5ba02f0bd4f2c0b09a403a41eded61a0), there is no `memcpy()` at all... Line 32 is blank
```
 29 _SYSCALL_HANDLER(spi_transceive, dev, config_p, tx_bufs, rx_bufs)
 30 {
 31         const struct spi_config *config = (const struct spi_config *)config_p;
 32 
 33         _SYSCALL_MEMORY_READ(config, sizeof(*config));
 34         _SYSCALL_DRIVER_SPI(dev, transceive);
```

I do not know what Shippable is smoking, but it has hallucinations.

**Comment by erwango** [comment]
recheck

**Comment by vaussard** [comment]
> Nice job! Hopefully this PR can be merged as soon as possible!

Hi @jli157, thank you. I hope also that this PR can be merged soon, presumably after the freeze window.

> recheck

Hi @erwango, I rebased on master and CI succeeded. It looks like all lights are green for merging.

**Comment by erwango** [comment]
@galak : Can you check then?

**Review by galak — CHANGES_REQUESTED** [review]
It seems like we need to update various stm32 board.yaml files to say we support PWM.

**Review by galak — APPROVED** [review]

**Comment by vaussard** [comment]
> It seems like we need to update various stm32 board.yaml files to say we support PWM.

Thank you @galak for the update.

**Comment by jli157** [comment]
Is anything still blocking this PR from being merged? I'm looking forward to using some new things from this PR. Thank you very much!

**Comment by vaussard** [comment]
I guess that new features should not be merged before Zephyr 1.12 is released.

**Comment by jli157** [comment]
Waiting for this one to be merged. Thanks!

**Pull request merged**
